### PR TITLE
GDPopt Enumerate

### PR DESCRIPTION
**Pull request by emma58**
## Summary/Motivation:

Introducing the slimmest, sleekest GDPopt algorithm yet! With no branching, not cuts, not even dual bounds, `gdpopt.enumerate` simply iterates over all the discrete solutions for your GDP and solves the corresponding (N)LPs! Simplify your life and enjoy for debugging and small problems, but keep in mind that $2^{100} \approx 1.27 \times 10^{30}$...

## Changes proposed in this PR:
- Vanilla implementation of enumeration algorithm: You can choose whether to enumerate over all discrete solutions or just the disjunctive realizations (meaning you solve MI(N)LPs if there are non-indicator discrete variables) using the `force_nlp_subproblem` option.
- Some tests for enumeration (but not a lot since, oddly enough, they can be costly)
- Many minor changes to the algorithm base class and the utilities around making and solving NLP subproblems, particularly as I realized all of the options that I assumed were defined in various places.

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/gdpopt/util.py [hunk 1]
```diff
@@ -187,14 +187,89 @@ def fix_discrete_var(var, val, config):
...
+                fix_discrete_var(subprob_var, val, self.config)
+                fixed_discrete.append("%s = %s" % (subprob_var.name, val))
+            self.config.logger.debug("Fixed the following integer variables: "
+                                     "%s" % ", ".join(fixed_discrete))
+
+        # Call the subproblem initialization callback
+        self.config.subproblem_initialization_method(self.True_disjuncts,
+                                                     self.boolean_var_values,
+                                                     self.discrete_var_values,
+                                                     self.subprob_util_block)
```
Comment (jsiirola) [inline pyomo/contrib/gdpopt/util.py:242]: As a general matter of practice, `__enter__()` should return something (usually `self`)

**Review by jsiirola — APPROVED** [review]
This looks pretty good.  One question around the definition of a context manager, but nothing to prevent merging.

File: pyomo/contrib/gdpopt/util.py (hunk 1 above)
Comment (emma58) [reply pyomo/contrib/gdpopt/util.py:243]: Oh, I did not know this. I just made these two return themselves, and also the suppress-infeasibility-warning one.

File: pyomo/contrib/gdpopt/enumerate.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                                                           util_block)
+
+        discrete_solns = list(self._discrete_solution_iterator(
+            subproblem_util_block.disjunction_list,
+            subproblem_util_block.non_indicator_boolean_variable_list,
+            subproblem_util_block.discrete_variable_list,
+            config))
+        self.num_discrete_solns = len(discrete_solns)
+        for soln in discrete_solns:
+            # We will interrupt based on time limit of iteration limit:
```
Comment (blnicho) [inline pyomo/contrib/gdpopt/enumerate.py:112]: Typo: "of" should be "or"

File: pyomo/gdp/tests/models.py [hunk 3]
```diff
@@ -329,7 +340,7 @@ def simpledisj2_rule(disjunct):
     return m
 
 def makeDisjunctionsOnIndexedBlock():
-    """Two disjunctions (one indexed an one not), each on a separate 
+    """Two disjunctions (one indexed an one not), each on a separate
     BlockData of an IndexedBlock of length 2
```
Comment (blnicho) [inline pyomo/gdp/tests/models.py:344]: Typo: "an" should be "and"

**Review by blnicho — APPROVED** [review]
I found two small typos in comments but otherwise this looks great

**Pull request merged**